Issue#33 first database migration

Arthurpanda opened the pull request:
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)


## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

ToDo

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

Valera3107 committed:
Add liquibase db migrations

Commit:
wip

Valera3107 committed:
Add some classes to module home-data-migration

Commit:
wip

Valera3107 committed:
Add some classes to module home-data-migration

Valera3107 committed:
Add some classes to module home-data-migration

Valera3107 committed:
Add some classes to module home-data-migration

Valera3107 committed:
Add some classes to module home-data-migration

Valera3107 committed:
Merge branch 'Issue#33-First-migration' into dev

# Conflicts:
#	home-data/src/main/resources/application.properties

Valera3107 committed:
Add fat jar file

Valera3107 committed:
Add fat jar file

Valera3107 committed:
Add fat jar file

Valera3107 committed:
Add fat jar file

Valera3107 committed:
Add fat jar file

Valera3107 committed:
Add profile called liquibase.migration

kh0ma commented on home-data-migration/pom.xml:
```diff
@@ -13,4 +13,79 @@
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-jdbc</artifactId>
```
Do we really need it here?

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/pom.xml:
```diff
@@ -13,4 +13,79 @@
+                    <plugin>
+                        <groupId>org.liquibase</groupId>
+                        <artifactId>liquibase-maven-plugin</artifactId>
+                        <version>3.6.3</version>
```
Move to property `liquibase.version`

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/pom.xml:
```diff
@@ -13,4 +13,79 @@
+        <dependency>
+            <groupId>org.liquibase</groupId>
+            <artifactId>liquibase-core</artifactId>
+            <version>4.0.0</version>
```
Reuse property `liquibase.version`

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/pom.xml:
```diff
@@ -13,4 +13,79 @@
+                <configuration>
+                    <finalName>runMigrations</finalName>
+                    <descriptorRefs>
+                        <descriptorRef>jar-with-dependencies</descriptorRef>
```
Can we produce only one artifact and remain the original name?

Arthurpanda replied:
Done

kh0ma commented on home-data-migration/pom.xml:
```diff
@@ -13,4 +13,79 @@
+                    <descriptorRefs>
+                        <descriptorRef>jar-with-dependencies</descriptorRef>
+                    </descriptorRefs>
+                    <archive>'
```
`'`

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/pom.xml:
```diff
@@ -13,4 +13,79 @@
+        </plugins>
+    </build>
 </project>
+
```
remove extra newline

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseRunner.java:
```diff
@@ -0,0 +1,15 @@
+package com.softserveinc.ita.homeproject.migrations;
```
Revisit package name.

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseRunner.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.Scanner;
+
+public class LiquibaseRunner {
+    public static void main(String[] args) {
```
Please consider using https://picocli.info/ or just parse args instead of promoting values from cli.

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseRunner.java:
```diff
@@ -0,0 +1,15 @@
+        Scanner scanner = new Scanner(System.in);
+        user = scanner.nextLine();
+        pass = scanner.nextLine();
+        LiquibaseUpdate liquibaseUpdate = new LiquibaseUpdate();
```
Why do we need this object? Can we provide some properties in it?

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseRunner.java:
```diff
@@ -0,0 +1,15 @@
+        user = scanner.nextLine();
+        pass = scanner.nextLine();
+        LiquibaseUpdate liquibaseUpdate = new LiquibaseUpdate();
+        liquibaseUpdate.runLiquibase(user, pass);
```
With a high probability URL of DB will vary on the environments. Please use the default if not present approach and give a possibility to specify it.

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+//    Get value from migration.properties
+    public String getFromProperties(String value) {
```
Please try to use util for it. e.g. Apache Commons, Google Guava

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+    public void runLiquibase(String user, String password) {
+        String url = getFromProperties(DB_HOST);
+        String path = getFromProperties(PATHNAME);
+        String changelogName = getFromProperties(CHANGELOG);
```
Can we inline it into path?

Arthurpanda replied:
Done

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+        String path = getFromProperties(PATHNAME);
+        String changelogName = getFromProperties(CHANGELOG);
+        try (Connection connection = DriverManager.getConnection(url, user, password)) {
+            DriverManager.registerDriver(new org.postgresql.Driver());
```
Please use import.

Arthurpanda replied:
Done

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+        String changelogName = getFromProperties(CHANGELOG);
+        try (Connection connection = DriverManager.getConnection(url, user, password)) {
+            DriverManager.registerDriver(new org.postgresql.Driver());
+            Database database = DatabaseFactory.getInstance().findCorrectDatabaseImplementation(new JdbcConnection(connection));
```
Can we omit it and provide `new JdbcConnection(connection)` directly into liquibase.

Arthurpanda replied:
Done

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+            FileSystemResourceAccessor fileSystemResourceAccessor = new FileSystemResourceAccessor();
+            ClassLoader classLoader = ClassLoader.getSystemClassLoader();
+            CommandLineResourceAccessor commandLineResourceAccessor = new CommandLineResourceAccessor(classLoader);
+            CompositeResourceAccessor compositeResourceAccessor = new CompositeResourceAccessor(fileSystemResourceAccessor, commandLineResourceAccessor);
```
Can we just use commandLineResourceAccessor?

Arthurpanda replied:
Yes, FileSystemResourceAccessor was deleted

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+            CommandLineResourceAccessor commandLineResourceAccessor = new CommandLineResourceAccessor(classLoader);
+            CompositeResourceAccessor compositeResourceAccessor = new CompositeResourceAccessor(fileSystemResourceAccessor, commandLineResourceAccessor);
+            try (Liquibase liquibase = new Liquibase(path+changelogName, compositeResourceAccessor, database)) {
+                liquibase.update("first_migration");
```
Can we get rid of `"first_migration"`?

Arthurpanda replied:
Yes, i`ve got it

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+            try (Liquibase liquibase = new Liquibase(path+changelogName, compositeResourceAccessor, database)) {
+                liquibase.update("first_migration");
+            } catch (LiquibaseException e) {
+                e.printStackTrace();
```
Please log the error and rethrow the error.

kh0ma commented on home-data-migration/src/main/java/com/softserveinc/ita/homeproject/migrations/LiquibaseUpdate.java:
```diff
@@ -0,0 +1,60 @@
+                e.printStackTrace();
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
Please log the error and rethrow the error.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/db.changelog-master.xml:
```diff
@@ -0,0 +1,12 @@
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+    <changeSet id="init-db" author="Artur" context="first_migration">
+        <tagDatabase tag="db_init"/>
```
Why do we need it?

kh0ma commented on home-data-migration/src/main/resources/db/changelog/db.changelog-master.xml:
```diff
@@ -0,0 +1,12 @@
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+    <changeSet id="init-db" author="Artur" context="first_migration">
```
author -> github login or email

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/v-1.0/001-changeset-users-table.xml:
```diff
@@ -0,0 +1,19 @@
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+    <changeSet id="001" author="Artur">
```
Can we provide a more valuable id? Instantly include the issue number in it? `id="001"`

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/v-1.0/001-changeset-users-table.xml:
```diff
@@ -0,0 +1,19 @@
+                      http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+
+    <changeSet id="001" author="Artur">
+        <sqlFile dbms="postgresql"
```
Please consider using liquibase dsl.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/v-1.0/001-create-users-table.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE if not exists public.users (
```
Do we need it? `if not exists`

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/v-1.0/001-create-users-table.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE if not exists public.users (
+                            id int not null  primary key,
```
Please revisit the data type. `int` -> `bigint`

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/v-1.0/001-create-users-table.sql:
```diff
@@ -0,0 +1,12 @@
+                            id int not null  primary key,
+                            first_name varchar (255)  not null,
+                            last_name varchar (255)  not null,
+                            login varchar (255) not null,
```
email

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/resources/db/changelog/v-1.0/001-create-users-table.sql:
```diff
@@ -0,0 +1,12 @@
+                            last_name varchar (255)  not null,
+                            login varchar (255) not null,
+                            password varchar (255)  not null,
+                            phone_number varchar (255) not null,
```
contact?

mykytam replied:
Fixed.

kh0ma commented on home-data-migration/src/main/resources/migration.properties:
```diff
@@ -0,0 +1,6 @@
+url = jdbc:postgresql://localhost:5432/test
+username=postgres
```
` = ` / `=`

mykytam replied:
Fixed.

mykytam committed:
Issue#33: First database migration
Fix package naming
Change SQL script
Add GitHub username to author xml files
Delete redundant additional spaces and lines

Valera3107 committed:
Add new inner class Builder to class LiquibaseRunner
Add new plugin to create only one .jar file
Change method for getting properties from file
Take out all versions of plugins and dependencies as a property value

Valera3107 committed:
Add scripts in xml format
Sql scripts format was deleted

Valera3107 committed:
Add property for heroku
Add opportunity to connect with full url, that includes user and password
Delete class Builder from LiquibaseUpdate

Valera3107 committed:
Add property for heroku
Add opportunity to connect with full url, that includes user and password
Delete class Builder from LiquibaseUpdate

kh0ma commented:
Moved to https://github.com/ita-social-projects/Home/pull/88
